BaseClasses.py

```diff
@@ -681,19 +681,57 @@ def can_reach_region(self, spot: str, player: int) -> bool:
-        locations = {location for location in locations if location.advancement and location not in self.events and
-                     not key_only or getattr(location.item, "locked_dungeon_item", False)}
-        while reachable_events:
-            reachable_events = {location for location in locations if location.can_reach(self)}
-            locations -= reachable_events
+        if key_only:
+            def event_filter(location: Location):
+                return location.item is not None and getattr(location.item, "locked_dungeon_item", False)
+        else:
+            def event_filter(location: Location):
+                return (location.advancement and location not in self.events
+                        or getattr(location.item, "locked_dungeon_item", False))
+        if locations is None:
+            # `self.multiworld.get_filled_locations(player)` is avoided because it first iterates into a list and also
+            # because `location.advancement` in `event_filter` also checks for `location.item is not None`.
+            events_per_player = {player: set(filter(event_filter, locations_dict.values()))
+                                 for player, locations_dict in self.multiworld.regions.location_cache.items()}
+        else:
+            # Can't iterate self.multiworld.player_ids because it doesn't include the extra player IDs used for item
+            # links, so iterate the keys of the location_cache instead.
+            events_per_player = {player: set() for player in self.multiworld.regions.location_cache.keys()}
+            for location in locations:
+                if event_filter(location):
+                    events_per_player[location.player].add(location)
+
+        # Remove empty sets to reduce the number of iterations required to iterate through `events_per_player.items()`.
+        for player, events in list(events_per_player.items()):
+            if not events:
+                del events_per_player[player]
+
+        # The first iteration must check the locations for all players because it is not known which players might have
+        # reachable locations.
+        players_to_check = set(self.multiworld.regions.location_cache.keys())
+        while players_to_check:
+            # Iterable of all locations belonging to the players to check.
+            loc_gen: Iterable[Location] = itertools.chain.from_iterable(
+                locations for player, locations in events_per_player.items() if player in players_to_check)
+            # Iterate reachable events into a list so that the sets in `events_per_player` can be modified while
+            # iterating through the reachable events.
+            reachable_events = [location for location in loc_gen if location.can_reach(self)]
+            # Clear the players to check to prepare for the next iteration.
+            players_to_check.clear()
             for event in reachable_events:
+                events_per_player[event.player].remove(event)
```

Comment:
I tried checking for the set being empty, after removing `event` from it, and then removing the set from `events_per_player` if it was empty, however this appeared to be slower.

I presume the cost of checking the set being empty for each reachable event was more than the cost of having to iterate even the empty sets in `events_per_player` for each outer loop.